[SG-74] Trial Initiation Component with Vertical Stepper

differsthecat opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Add Trial Initiation page and vertical stepper component.

## Code changes

- **Add** Trial Initiation Module and Component
- **Add** Vertical Stepper component
    - Extends [Angular CDK stepper ](https://material.angular.io/cdk/stepper/api)
    - Uses Tailwind for all styles
- **Update** `apps/web/src/app/oss-routing.module.ts` to add new trial route

## Screenshots

Demo:
https://user-images.githubusercontent.com/8926729/174153198-6f7d7515-e7f8-4359-a1af-a9f019f65af7.mov

Light Mode:
![Screen Shot 2022-06-20 at 3 47 06 PM](https://user-images.githubusercontent.com/8926729/174669122-b25b66a2-04d8-447a-9d6f-5bd6a043c9ef.png)

Dark mode:
![Screen Shot 2022-06-20 at 3 46 42 PM](https://user-images.githubusercontent.com/8926729/174669098-dcb6d019-f8a8-4979-8d9a-d20fb59dfe2b.png)


## Before you submit

<!-- (mark with an `X`) -->

```
- [x] I have checked for **linting** errors (`npm run lint`) (required)
- [ ] I have added **unit tests** where it makes sense to do so (encouraged but not required)
- [ ] This change requires a **documentation update** (notify the documentation team)
- [ ] This change has particular **deployment requirements** (notify the DevOps team)
```

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,102 @@
+  <header class="tw-bg-[#175ddc] tw-text-white tw-h-16">
+    <div class="">
+      <div
+        class="tw-bg-[#175ddc] tw-content-[''] tw-h-64 tw-left-0 tw-absolute tw--top-20 tw--skew-y-3 tw-w-full tw--z-10"
```
We should avoid arbitary values per https://contributing.bitwarden.com/code-style/tailwind/.

differsthecat replied:
Got it! What should I do in the case where I need our primary blue color like in this property regardless of light/dark mode?

danielleflinn replied:
I would recommend using one of the following:
-  `background-alt2` (shows primary in light theme but a darker color in dark theme). 
- 'primary-500'; this will require an update to increase contrast of the logo and text on the header for dark theme.

@jtouchstonebw if you don't think either of the above will work for the page's design we can look at adding another variable to the tailwind config that is `#175ddc` in both light and dark.

jtouchstonebw replied:
Agree with @danielleflinn s suggestion.

danielleflinn replied:
`background-alt2` is what @jtouchstonebw and I aligned on

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.ts:
```diff
@@ -0,0 +1,35 @@
+      if (qParams.email != null && qParams.email.indexOf("@") > -1) {
+        this.email = qParams.email;
+      }
+      switch (qParams.layout) {
+        case "families":
+          this.layout = "families";
+          break;
+        case "enterprise":
+        case "enterprise1":
+        case "enterprise2":
+          this.layout = "enterprise";
+          break;
+        default:
+          this.layout = "teams";
+          break;
+      }
```
I always disliked the way the register component was built since it's quite a nightmare to handle. What are your thoughts of either making different components for the different layouts, that all use the same right side register component? Or make the content data driven from json.

differsthecat replied:
Oh that's a good idea! I'll play around with that

micahblut replied:
We have SG-65 for making registration into a reusable component. Does that conflict with this feedback at all?

Hinton replied:
I would say that they compliment each other. We'll probably want to figure out a way to do so in this task and re-use the same method in SG-65.

Hinton commented on apps/web/src/app/modules/trial-initiation/vertical-stepper/vertical-stepper.component.html:
```diff
@@ -0,0 +1,67 @@
+<div>
+  <ul class="tw-flex tw-list-none tw-flex-col tw-flex-wrap tw-p-5">
+    <li *ngFor="let step of steps; let i = index; let isLast = last">
+      <div class="tw-w-74 tw-m-2.5 tw-text-center tw-h-16">
```
Do we need to use widths everywhere? This makes the components quite fragile compared to having them expand to the parents width.

differsthecat replied:
Probably not needed, when I translated the css to tailwind I did more of a 1 to 1 thing. I'll take a better look at this and see if I can change this

differsthecat replied:
I was able to remove this one here, but I needed to keep the widths set on the `button` and the `span`s for the circles below it.

addisonbeck commented on apps/web/src/app/modules/loose-components.module.ts:
```diff
@@ -161,14 +161,21 @@ import { FolderAddEditComponent } from "../vault/folder-add-edit.component";
+    VaultFilterModule,
+    OrganizationBadgeModule,
+    PipesModule,
+    TrialInitiationModule,
```
Do you need this? I've been questioning the reason this imports array exists at all, but TrialInitiationModule specifically shouldn't be relevant to these components because that's net new work, correct?

differsthecat replied:
Just checked and we don't, I was in a bit of auto-pilot and mimicked the VaultFilter stuff

differsthecat commented on apps/web/src/app/modules/trial-initiation/vertical-stepper/vertical-stepper.module.ts:
```diff
@@ -0,0 +1,12 @@
+  declarations: [VerticalStepperComponent, VerticalStepComponent],
+  exports: [VerticalStepperComponent, VerticalStepComponent],
+})
+export class VerticalStepperModule {}
```
Yes, absolutely!

Hinton reviewed (changes requested):
This looks like a good start, however there are a few areas of concerns that I think we need to address before this can get merged.

I apologize for the huge number of comments, but I think we need to spend some additional time getting the tailwind styling right on this. I wonder if the direct migration from bootstrap to tailwind has resulted in a quite odd way of writing tailwind styles. As it's currently written it will be quite difficult to maintain, particularly if we ever wish to add responsive design to it.

I've highlighted a few tailwind classes that are either unnecessary, or confusing. I also believe the use of `!` is a sign that something is off, we shouldn't need to use that many `!important` unless we're struggling with bootstrap base styling.

We seem to be using content projection for some reason that I can't quite figure out. I don't believe it's necessary for this functionality and it should be avoided unless there are no better alternative since it's a quite advanced angular functionality.

We should also consider if we want the trial component to drag in the different marketing components, or flip the relationship which would provide more customization in how marketing can structure the pages. Particularly we should avoid ever getting in a scenario where we have two variables that controls the styling, or if there is ever a want/need to put different content outside the left side. (Which I already suspect their is since we use different subheaders)

differsthecat commented:
> This looks like a good start, however there are a few areas of concerns that I think we need to address before this can get merged.
> 
> I apologize for the huge number of comments, but I think we need to spend some additional time getting the tailwind styling right on this. I wonder if the direct migration from bootstrap to tailwind has resulted in a quite odd way of writing tailwind styles. As it's currently written it will be quite difficult to maintain, particularly if we ever wish to add responsive design to it.
> 
> I've highlighted a few tailwind classes that are either unnecessary, or confusing. I also believe the use of `!` is a sign that something is off, we shouldn't need to use that many `!important` unless we're struggling with bootstrap base styling.
> 
> We seem to be using content projection for some reason that I can't quite figure out. I don't believe it's necessary for this functionality and it should be avoided unless there are no better alternative since it's a quite advanced angular functionality.
> 
> We should also consider if we want the trial component to drag in the different marketing components, or flip the relationship which would provide more customization in how marketing can structure the pages. Particularly we should avoid ever getting in a scenario where we have two variables that controls the styling, or if there is ever a want/need to put different content outside the left side. (Which I already suspect their is since we use different subheaders)

Thank you @Hinton for the feedback!
You are correct, I did more of a one-to-one conversion of the css for the register component to this new component, so that combined with my inexperience with tailwind has definitely created some oddities.

As for the content projection for the step label, I took another look and was able to remove it. I probably ran into some issue when I originally did that, and never went back to take another look at it.

I tried to keep the variables controlling the styling that marketing is using because they want to do something with the `layout` parameter (or are doings something with it) for gathering/understanding metrics as far as I knew. I'll work with them more on this, I figured it was safe to leave to them to use as they please here.

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,76 @@
+  <header class="tw-bg-background-alt2 tw-text-white tw-h-16">
+    <div class="">
+      <div
+        class="tw-bg-background-alt2 tw-content-[''] tw-h-64 tw-left-0 tw-absolute tw--top-20 tw--skew-y-3 tw-w-full tw--z-10"
```
`tw-content-['']` was a carry over from the old css that I translated over. Removed that, and implemented your other suggestions

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,76 @@
+      <div
+        class="tw-bg-background-alt2 tw-content-[''] tw-h-64 tw-left-0 tw-absolute tw--top-20 tw--skew-y-3 tw-w-full tw--z-10"
+      >
+        <div class="tw-flex-wrap"></div>
```
Nope, removed!

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,76 @@
+      </div>
+    </div>
+    <div>
+      <div class="tw-w-[650px] tw-block tw-pt-56">
```
I couldn't figure out how to get the width to be larger than the component itself with percentages as tailwind's largest value is `96` which translates to 384px. I made this a hardcoded percentage instead of pixels, but I do not think that is what you were meaning. Do you think this will resolve once we start adding the actual content to the stepper?

Hinton replied:
Tailwind has percentage based widths that should be used for columns. The pixel based are more when you need a sidebar in a fixed with which is why they only go to `w-96`.

We should either set the left side to a fixed value like `w-96` and have the register occupy the rest. Or set both to percentage based, `w-4/12`, `w-8/12` or

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,76 @@
+    </div>
+  </header>
+
+  <div class="tw-flex tw-justify-center tw--mt-12">
```
This one was complicated by a couple of things. I wanted the logo go over the skewed section without being skewed itself, and staying in line with the left hand content of the page as well. I'm definitely open to a better way to do this, this seemed to be the most straightforward way to do this.

Hinton replied:
In this scenario I would probably have created the background in the header using a separate div with position absolute. And then have the texts in the header as well which wouldn't care about the background since it's a sibling with position absolute.

```
<header class="tw-h-64 tw-w-full">
  <div background></div>
  <img/>
  <h1>Page Header</h1>
</header>

differsthecat replied:
I tried playing around with this, but I found it got pretty complicated to get the logo and the text to stay in line with the left side content depending on how the page is sized. I'd like to leave it as is if that is ok with you.

Hinton replied:
Do we have any plans for how the responsive implementation of the page will behave? Will the text be above the form? If so then it's probably fine but if we plan on having the form before the text then it won't work with the current structure.

differsthecat replied:
The current plan as far as I'm aware is to always have the text before the form (on the left) but feel free to correct me if I am wrong here @jtouchstonebw

jtouchstonebw replied:
I was following what we had at the time for the web vault, as it was not responsive. If making things responsive is within the scope of this work, then having the text either on the left or above the form is fine.

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,76 @@
+      <div>
+        <img
+          alt="Bitwarden"
+          class="tw-h-[57px] tw-w-72 tw-mt-2.5 tw-mb-0 tw-mx-0"
```
Good to know, I thought their widths would remain the values defined in the docs

Hinton replied:
They should remain mostly stable, but since they are customizeable there is no guarantee they won't change in a Tailwind upgrade, or if we decide to slightly tweak it.

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,76 @@
+<div class="">
+  <header class="tw-bg-background-alt2 tw-text-white tw-h-16">
```
I found it in the tailwind docs themselves, but looking at our colors we don't have it. Removed

differsthecat commented on apps/web/src/app/modules/vertical-stepper/vertical-step.component.html:
```diff
@@ -0,0 +1,7 @@
+<ng-template>
```
So that the content of the step can be rendered by the stepper.

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,76 @@
+          </div>
+          <app-vertical-stepper #stepper linear>
+            <!-- Content is for demo purposes. Replace with form components for each step-->
+            <app-vertical-step #step1 label="Create Account" [editable]="false">
```
Oh, I see what you mean. I was using `#stepper` one for the `stepper.reset()` function for the demo, but I can remove that. The other ones aren't needed either, so this is fixed!

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.ts:
```diff
@@ -0,0 +1,29 @@
+      }
+      if (qParams.layout) {
+        // For marketing use
+      }
```
Removed for now, I'll get with marketing and coordinate with them on what they need when they are ready to update this content 👍

differsthecat commented on apps/web/src/app/modules/vertical-stepper/vertical-stepper.component.html:
```diff
@@ -0,0 +1,67 @@
+          class="!tw-w-full tw-h-full tw-flex tw-border-none tw-cursor-pointer tw-bg-transparent tw-items-center tw-flex-row tw-max !tw-cursor-default"
+          [ngClass]="{
+            'hover:tw-bg-secondary-100 !tw-cursor-pointer': !isStepDisabled(i) && step.editable
+          }"
```
I updated this to use `[disabled]` for the button, but left the hover class in to align with the design specs

Hinton replied:
We should be able to use `disabled:hover:tw-bg-secondary-100`?

differsthecat replied:
I just tried it and when I remove the disabled attribute and placed it in `ngClass` instead, the cursor indicates disabled buttons are clickable when they are not

Hinton commented on apps/web/src/app/modules/trial-initiation/enterprise-content.component.html:
```diff
@@ -0,0 +1,12 @@
+<div class="tw-max-w-3xl tw-min-w-fit">
+  <h1 class="!tw-text-[#ffffff]">You've chosen Bitwarden for Enterprise</h1>
```
@danielleflinn Seems we have a couple of cases where we always want to use a white text. Should we add it to the colors?

Hinton replied:
@differsthecat do we need the `!` here?

differsthecat replied:
Yes we do on these, it presents as `tw-text-main` (with the text being black in light mode) if it's not marked important

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,63 @@
+              Start your 7-Day free trial of Bitwarden for {{ org }}
+            </h2>
+          </div>
+          <app-vertical-stepper linear>
```
We do as a property for the base class `CdkStepper` to enforce the linear behavior.

differsthecat commented on apps/web/src/app/modules/vertical-stepper/vertical-step-content.component.html:
```diff
@@ -0,0 +1,47 @@
+  <button
+    (click)="selectStep()"
+    [disabled]="disabled"
+    class="tw-w-full tw-h-full tw-flex tw-border-none tw-cursor-pointer tw-bg-transparent tw-items-center tw-flex-row tw-max"
```
Just checked again and the width is needed, but height is not needed and I removed `tw-max`.

Without the width, the button isn't the full width of it's parent like we want: 
![Screen Shot 2022-06-23 at 8 08 10 AM](https://user-images.githubusercontent.com/8926729/175296757-7b744369-30ea-480c-b9ed-1cdeed5ffff1.png)

differsthecat commented on apps/web/src/app/modules/vertical-stepper/vertical-step-content.component.html:
```diff
@@ -0,0 +1,47 @@
+      [ngClass]="{
+        '!tw-text-contrast !tw-bg-primary-500': selected,
+        '!tw-text-main !tw-bg-secondary-300': !selected && !disabled && step.editable,
+        '!tw-text-muted !tw-bg-transparent': disabled
```
Nope, just checked and they are not necessary!

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,63 @@
+        <app-families-content *ngIf="org === 'families'"></app-families-content>
+      </div>
+    </div>
+    <div>
```
Yes, it's to display as a block with its sibling div. I had the classes a little wonky so I updated them a bit to make it clearer

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,63 @@
+<div>
```
Removed!

differsthecat commented on apps/web/src/app/modules/trial-initiation/enterprise-content.component.html:
```diff
@@ -0,0 +1,12 @@
+<div class="tw-max-w-3xl tw-min-w-fit">
```
I checked and found they aren't needed!

differsthecat commented on apps/web/src/app/modules/vertical-stepper/vertical-step-content.component.html:
```diff
@@ -0,0 +1,47 @@
+    >
+      <i class="bwi bwi-fw bwi-check tw-p-1" aria-hidden="true"></i>
+    </span>
+    <ng-container>
```
nope, removed!

differsthecat commented:
@Hinton Here is a summary of changes I have made:

- Removed all unnecessary instances of `!` important modifiers
- Removed unnecessary divs and cleaned up styles for ones that are needed
- other style changes.

I commented back on your feedback with my reasoning for the rest of the items.
Thank you again!

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -1,62 +1,60 @@
-          </app-vertical-stepper>
+  </div>
+  <div class="tw-block">
+    <div class="tw-w-[150%] tw-pt-56">
```
We still need to figure out this. The normal way of implementing two columns is to have:

```html
<div class="flex">
	<div class="w-1/3">
		Left column
  </div>
  <div class="w-2/3">
    Right column
  </div>
</div>
```

This has the added benefit of quite easily making the page responsive by simply removing the flex in the wrapper, and the left column gets displayed above the right.

differsthecat replied:
That worked out perfectly and is better, thank you!

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,61 @@
+  <div class="tw-block">
+    <div class="tw-w-[150%] tw-pt-56">
+      <div class="tw-rounded tw-border tw-border-solid tw-bg-background tw-border-secondary-300">
+        <div class="tw-h-12 tw-flex tw-items-center tw-rounded-t tw-bg-secondary-100 tw-w-full">
```
This div seems to only have a single child, do we need it to be flex vs block?

differsthecat replied:
Yes, updated

differsthecat commented on apps/web/src/app/modules/vertical-stepper/vertical-step-content.component.html:
```diff
@@ -2,46 +2,44 @@
   >
     <span
-      class="tw-rounded-full tw-flex-row tw-font-bold tw-leading-9 tw-mr-3.5 tw-items-center tw-w-9"
+      class="tw-rounded-full tw-flex-row tw-font-bold tw-leading-9 tw-mr-3.5 tw-w-9"
```
removed

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,61 @@
+    <div>
+      <img
+        alt="Bitwarden"
+        class="tw-h-[57px] tw-w-72 tw-mt-2.5 tw-mb-0 tw-mx-0"
```
Used style directly for this, I put them in line if that is what we are going for here, but I can make this a class instead in our styles to get rid of that (that is what I understood from your comment).

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -9,7 +9,8 @@
         alt="Bitwarden"
-        class="trial-logo tw-mt-2.5 tw-mb-0 tw-mx-0"
+        style="height: 50px; width: 335px"
+        class="tw-mt-2.5 tw-mb-0 tw-mx-0"
```
Isn't the margin 0 on images by default?

differsthecat replied:
Just checked and it is, fixed 👍

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,62 @@
+
+<div class="tw-flex tw-justify-center tw--mt-12">
+  <div class="tw-w-1/3">
+    <div>
```
This seems to be unnecessary during my testings.

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,62 @@
+<header class="tw-bg-background-alt2 tw-h-20">
+  <div>
```
Also seems unnecessary.

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,62 @@
+      <app-families-content *ngIf="org === 'families'"></app-families-content>
+    </div>
+  </div>
+  <div class="tw-w-1/3">
```
This should probably be `tw-w-2/3`, since we currently don't consume the full width. (or have both be `tw-w-1/2`

Hinton reviewed (changes requested):
There are two additional issues I've discovered. The tiling of the background breaks on large monitors.

![image](https://user-images.githubusercontent.com/137855/176022527-10b9d444-bfd0-4317-9aa3-8ad2583c0057.png)

We also need to fix the responsiveness of the page. Currently it always occupies 2/3 of the page, this doesn't work with either large monitors or small monitors. In case we don't want to provide support for small displays we need to set a min-width.

For larger monitors we need to provide a max-with on a container. This "container" can be the `tw-flex` container you already have, and probably needs padding + `tw-mx-auto` to be centered. https://tailwindui.com/components/application-ui/application-shells/stacked shows an example of how a container can be setup.


We need to figure out what to do with the responsiveness. We should give the container a `min-width` to prevent the page from breaking. (And probably minimize the page margins)

differsthecat commented:
> There are two additional issues I've discovered. The tiling of the background breaks on large monitors.
> 
> We also need to fix the responsiveness of the page. Currently it always occupies 2/3 of the page, this doesn't work with either large monitors or small monitors. In case we don't want to provide support for small displays we need to set a min-width.
> 
> For larger monitors we need to provide a max-with on a container. This "container" can be the `tw-flex` container you already have, and probably needs padding + `tw-mx-auto` for it to be centered. https://tailwindui.com/components/application-ui/application-shells/stacked shows an example of how a container can be setup.
> 
> Something like
> 
> ```
> <div class="tw-flex tw-justify-center tw--mt-12 tw-max-w-7xl tw-min-w-4xl tw-mx-auto tw-px-4">
> ```
> 
> should probably work. It should resize between `1280px` and `896px`.

Thank you @Hinton, those were some good catches!
I fixed the tiling of the background and implemented your suggested changes for the responsiveness. 
I did end up going with an arbitrary value for the max width of 1280px as `tw-max-w-7xl` was evaluating as 1120px, which seemed to be a bit too small. Here is a demo of me zooming in and out (to simulate larger and smaller monitors) and changing the size of the window as well:

https://user-images.githubusercontent.com/8926729/176041683-2317f44a-fb95-4f7e-afe5-556fe2e746ce.mov

Hinton commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,59 @@
+<header class="tw-bg-background-alt2 tw-h-10">
+  <div class="tw-bg-background-alt2 tw-h-48 tw-absolute tw--skew-y-3 tw-w-full tw--z-10"></div>
+  <div class="tw-w-1/2 tw-bg-background-alt2 tw-h-40"></div>
+</header>
+
+<div class="tw-flex tw-justify-center tw--mt-12 tw-max-w-[1280px] tw-min-w-4xl tw-mx-auto tw-px-4">
```
We can replace the header with this. It sets a height of 96, puts a negative margin of half that, and skews it. From what I can tell it's identical to the previous behavior.

```suggestion
<div class="tw-bg-background-alt2 tw-h-96 tw--mt-48 tw-absolute tw--skew-y-3 tw-w-full tw--z-10"></div>

<div class="tw-flex tw-justify-center tw-max-w-screen-xl tw-min-w-4xl tw-mx-auto tw-px-4">
```

differsthecat replied:
Just checked and this is good! 👍

differsthecat commented on apps/web/src/app/modules/trial-initiation/trial-initiation.component.html:
```diff
@@ -0,0 +1,59 @@
+  <div class="tw-w-1/2 tw-bg-background-alt2 tw-h-40"></div>
+</header>
+
+<div class="tw-flex tw-justify-center tw--mt-12 tw-max-w-[1280px] tw-min-w-4xl tw-mx-auto tw-px-4">
```
Resolved by below changes (`tw-max-w-screen-xl` was the ticket)

Hinton commented:
Linter is unhappy :/
